src/resmom/mom_hook_func.c

```diff
@@ -886,16 +885,18 @@ run_hook(hook *phook, unsigned int event_type, mom_hook_input_t *hook_input,
 				log_err(errno, __func__, msg_err_malloc);
 				return (-1);
 			}
-		 	if (php)
+		 	if (php) {
 				ptask->wt_parm2 = (void *)php;
+				if (php->hook_input && php->hook_input->pjob)
+					php->hook_input->pjob->ji_bg_hook_task = ptask;
```

Comment:
This assumes there is only one end hook.  There can be multiple hooks with the same hook event.  How do they get run in the background?  One after the other?  If we continue to run and purge the job after the first one, how do we run the subsequent ones?  Also, this is assuming there is just one task.  Will this single task pointer work with multiple end hooks?

Reply:
mom_process_background_hooks() handles running subsequent end job hooks.
After the first background hook runs to completion - post_run_hook() is run.
post_hook_run() will create a task (with func = mom_process_background_hooks) to check and run next hook script.
So, there is actually only one task at a time, subsequent end job hooks are run by creating tasks.

With this fix, we do not run the subsequent endjob hooks.

Reply:
I think it is a bad idea to not run all of the end hooks.  It would be possible for the site to add several end hooks with order < 100, and then the cgroups hook might not run and the cgroups won't be torn down.  Even if we document this, we're telling the sites that if they have multiple end hooks, they might not all run.  To me this breaks our hook infrastructure.  How hard would it be to keep the job around and then purge it in the last call of post_run_hook()?
Hmm here's another wrinkle.  What happens if we run one end hook, and  by the time it finishes, the job is requeued and run again on this node.  Will post_run_hook() find the new version of the job and run more end hooks?

Reply:
@bhroam : you've made a good point! True, if there are multiple end hooks, the pbs_cgroups end hook might not execute, and cgroups won't get cleaned up. That would be bad. If there was only one end hook, which can be checked by doing:

GET_NEXT(pjob->ji_bg_hook_task->wt_parm1->hi_execjob_end_hooks) == NULL
then the current fix scheme would work. But if there are other execjob_end hooks, we might just let the remaining execjob_end hooks execute in blocking mode. This means mom would block until all the execjob_end hooks execute on behalf of the job, before proceeding to free the job totally.
Now trying to keep the job around and then purge in the last call of post_run_hook() seemed tricky. We would need to go through the mom code looking for the job_purge_mom() calls. @prakashcv13 : try to see the actual sequence that happens when a job is interrupted (like the prologue hook rejecting request), and then requeued/rerun. I just don't know where in mom code that particular job_purge_mom() call happens, and if it's safe to not let it happen...

Reply:
@prakashcv13 : basically, we'd like to achieve what happens when a normal,  non-interrupted job is executing end hooks, where it would linger around in Exiting state, until the last end hook executes. So if a job is interrupted, the job would still stay around until the final end hook executes. This means requeueing and reruning would get delayed as well.

Reply:
the sequence of events goes like this:
1) MS joins job with sisters
2) sisters reject the prologue event
3) MS aborts the job - runs the execjob_abort hooks including the cgroup hook which deletes the cgroups related to the job
4) MS sends abort request to sisters - sisters also run execjob_abort hooks
5) sisters purge the job
6) MS sends job obit to server
7) server reruns the job - sends delete job request from on_job_rerun
8) MS relays the request to sisters
9) MS runs endjob hooks
10) sisters do not find the job as it was purged in step   above - return error PBSE_JOBEXIST
11) MS rejects the delete job request from the server (chk_del_job) BUT adds the job the list of dead jobs
12) MS on itself purges the job in it's main loop
13) server on getting the reject reply (15059) sends a discard job request
14) MS does not find the job as it is already purged AND sends a success reply
15) server requeues the job
16) scheduler runs the job on the same nodes
17) post_run_hook() runs and MS crashes

Reply:
@bayucan, @bhroam - if the concern is whether the cgroups related to the job are getting deleted - they are getting deleted in the abort job event and we can go ahead with the fix, this is a very corner case according to me.
if the concern is that *all* endjob hooks should run - it is a race. Please let me know what do you guys think.

Reply:
Interesting, so the abort hook and the end hook runs?  I thought the purpose of the abort hook was to run it and not the end hook.  There were cases where the end hook wasn't running.  We didn't want to run the end hook there because existing end hooks might be run when the customer didn't expect them.  This is really weird because the abort hooks run everywhere, but only the end hook runs on MS.  I really dislike the fact that all the end hooks won't run.  The customer is telling us that they have N hooks and expect them to all run at the end of a job.  In this case, only some of them will run, and the number won't be predictable.  If the hooks are fast, you might run several, but if the hooks take a while, you'll run less.  Maybe the answer is to make the abort hook or the end hook run, but not both.  That will fix this problem since the abort hook will run.

Reply:
@prakashcv13 : Thank you for enumerating the steps that mom and server go through. This is really helpful. 
it is true that pbs_cgroups hook guards possible failures by doing clean up in the execjob_abort, in case execjob_end hook is not called. So the concern for the pbs_cgroups hook side is alleviated. 
However, we should explore if there's another way so that all execjob_end hooks are given a chance to execute before a job is rerun. Looking at the 4 steps below:
11. MS rejects the delete job request from the server (chk_del_job) BUT adds the job the list of dead jobs
12. MS on itself purges the job in it's main loop
13. server on getting the reject reply (15059) sends a discard job request
14. MS does not find the job as it is already purged AND sends a success reply

I believe the 2 places in mom that looks into mom_deadjobs list is as follows:
```
             i = 0;
                while ((pjob = (job *)GET_NEXT(mom_deadjobs)) != NULL) {
                        /* sometimes this purge is happening earlier than
                         * IS_DISCARD_JOB, which then does not get the pjob
                         * pointer to call kill_job().
                         *
                         * Fixed by adding a kill_job here, which should be
                         * no harm anyway.
                         */
                        (void)kill_job(pjob, SIGKILL);
                        job_purge_mom(pjob);
                        ++i;
                }
```
...
   ```
   while ((pjob = (job *)GET_NEXT(mom_deadjobs)) != NULL) {
                job_purge_mom(pjob);
        }
```

`
What if we prevent the call to job_purge_mom() from taking place in the loop body, until all end hooks have been processed. Anyway, the job would still be in the mom_deadjobs() list which mom's main loop would continue to check.

Now in step 14 above, perhaps MS should not only check if it does not find the job and reply success, but check also if job is NOT in mom_deadjobs, before replying success. Maybe that would prevent server from requesting the job to be requeued and prevent it from rerunning as job would still be in a R state (or actually E state). Eventually, job would disappear in mom_deadjobs list, and mom and server would be in sync and job goes back to Q state...